File: Lv.2/올바른 괄호/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/12909
class Stack():
    def __init__(self):
        self.s = []
        
    def push(self, item):
        if self.s and item == ')' and self.s[-1] == '(':
            self.s.pop()
        else:
            self.s.append(item)
            
    def is_empty(self):
        return True if not self.s else False

        
def solution(s):
    stack = Stack()
    
    for p in s:
        stack.push(p)
    
    return stack.is_empty()

File: Lv.2/올바른 괄호/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/12909?language=javascript
class Stack {
  s = [];
  Stack() {}

  push = (item) => {
    if (this.s.length && item === ")" && this.s[this.s.length - 1] == "(") {
      this.s.pop();
    } else {
      this.s.push(item);
    }
  };

  is_empty = () => {
    return this.s.length ? false : true;
  };
}

function solution(s) {
  const stack = new Stack();

  [...s].forEach((element) => {
    stack.push(element);
  });

  return stack.is_empty();
}
